ably/realtime_channel.go

```diff
@@ -2,130 +2,342 @@ package ably
 
 import (
 	"errors"
-	"fmt"
+	"sort"
 	"sync"
 
 	"github.com/ably/ably-go/ably/proto"
 )
 
-func NewRealtimeChannel(name string, client *RealtimeClient) *RealtimeChannel {
-	return &RealtimeChannel{
-		Name:      name,
-		client:    client,
-		listeners: make(map[string]map[chan *proto.Message]struct{}),
+type chanSlice []*RealtimeChannel
+
+func (ch chanSlice) Len() int           { return len(ch) }
+func (ch chanSlice) Less(i, j int) bool { return ch[i].Name < ch[j].Name }
+func (ch chanSlice) Swap(i, j int)      { ch[i], ch[j] = ch[j], ch[i] }
+func (ch chanSlice) Sort()              { sort.Sort(ch) }
+
+// Channels is a goroutine-safe container for realtime channels that allows
+// for creating, deleting and iterating over existing channels.
+type Channels struct {
+	mtx    sync.Mutex
+	client *RealtimeClient
+	chans  map[string]*RealtimeChannel
+}
+
+func newChannels(client *RealtimeClient) *Channels {
+	return &Channels{
+		client: client,
+		chans:  make(map[string]*RealtimeChannel),
+	}
+}
+
+// Get looks up a channel given by the name and creates it if it does not exist
+// already.
+//
+// It is safe to call Get from multiple goroutines - a single channel is
+// guaranteed to be created only once for multiple calls to Get from different
+// goroutines.
+func (ch *Channels) Get(name string) *RealtimeChannel {
+	ch.mtx.Lock()
+	c, ok := ch.chans[name]
+	if !ok {
+		c = newRealtimeChannel(name, ch.client)
+		ch.chans[name] = c
 	}
+	ch.mtx.Unlock()
+	return c
 }
 
-type ChanState int
+// All returns a list of created channels.
+//
+// It is safe to call All from multiple goroutines, however there's no guarantee
+// the returned list would not list a channel that was already released from
+// different goroutine.
+//
+// The returned list is sorted by channel names.
+func (ch *Channels) All() []*RealtimeChannel {
+	ch.mtx.Lock()
+	chans := make([]*RealtimeChannel, 0, len(ch.chans))
+	for _, c := range ch.chans {
+		chans = append(chans, c)
+	}
+	ch.mtx.Unlock()
+	chanSlice(chans).Sort()
+	return chans
 
-const (
-	ChanStateInitialized ChanState = iota
-	ChanStateAttaching
-	ChanStateAttached
-	ChanStateDetaching
-	ChanStateDetached
-	ChanStateFailed
-)
+}
+
+// Release closes a channel looked up by the name.
+//
+// It is safe to call Release from multiple goroutines - if a channel happened
+// to be already concurrently released, the method is a nop.
+func (ch *Channels) Release(name string) error {
+	ch.mtx.Lock()
+	defer ch.mtx.Unlock()
+	if c, ok := ch.chans[name]; ok {
+		delete(ch.chans, name)
+		return c.Close()
+	}
+	return nil
+}
 
+// RealtimeChannel represents a single named message channel.
+//
+// The name of the channel is read-only - assigning new name to it does not
+// have any effect.
 type RealtimeChannel struct {
-	Name string
+	Name string // read-only name of the channel
```

Comment:
Rather than this comment, how about not exporting it and adding a `Name()` method?